File: prime_seive.py
def count_primes(N: int) -> int:
    """
    Counts the number of prime numbers less than or equal to N using the Sieve of Eratosthenes.

    @param N: The upper limit for checking primes.
    @return: The number of primes less than or equal to N.
    """
    if N < 2:
        return 0
    
    is_prime = [True] * (N + 1)
    is_prime[0], is_prime[1] = False, False
    
    p = 2
    while p * p <= N:
        if is_prime[p]:
            for i in range(p * p, N + 1, p):
                is_prime[i] = False
        p += 1
    
    prime_count = sum(is_prime)
    
    return prime_count

if __name__ == "__main__":

    N = int(input().strip())
    
    result = count_primes(N)

    print(result)

File: prime_seive.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int count_primes(int N) {
    if (N < 2) return 0;

    bool *is_prime = (bool *)malloc((N + 1) * sizeof(bool));
    for (int i = 0; i <= N; i++) is_prime[i] = true;
    is_prime[0] = is_prime[1] = false;

    for (int p = 2; p * p <= N; p++) {
        if (is_prime[p]) {
            for (int i = p * p; i <= N; i += p) {
                is_prime[i] = false;
            }
        }
    }

    int prime_count = 0;
    for (int i = 2; i <= N; i++) {
        if (is_prime[i]) prime_count++;
    }

    free(is_prime);
    return prime_count;
}

int main() {
    int N;
    scanf("%d", &N);
    printf("%d\n", count_primes(N));
    return 0;
}
